charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            self.logger.warning("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _check_container_connection(self) -> None:
+        """Check if connection can be made with container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
+
+    def _check_status(self) -> None:
+        """Check status of workload and set status accordingly."""
+        self._check_leader()
+        container = self.unit.get_container(self._container_name)
+        if container:
+            try:
+                check = container.get_check("tensorboard-controller-up")
+            except ModelError as error:
+                raise GenericCharmRuntimeError(
+                    "Failed to run health check on workload container"
+                ) from error
+            if check.status != CheckStatus.UP:
+                self.logger.error(
+                    f"Container {self._container_name} failed health check. It will be restarted."
+                )
+                raise ErrorWithStatus("Workload failed health check", MaintenanceStatus)
+            else:
+                self.model.unit.status = ActiveStatus()
+
+    def _on_install(self, _) -> None:
+        """Handle install event."""
+        # deploy K8s resources to speed up deployment
+        self._apply_k8s_resources()
+
+    def _on_upgrade(self, _) -> None:
+        """Handle upgrade event."""
+        # force conflict resolution in K8s resources update
+        self._on_event(_, force_conflicts=True)
+
+    def _on_remove(self, _) -> None:
+        """Handle remove event."""
+        delete_error = None
+        self.unit.status = MaintenanceStatus("Removing K8s resources")
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        crd_resources_manifests = self.crd_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as error:
+            # do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Failed to delete CRD resources, with error: {error}")
+                delete_error = error
+        try:
+            delete_many(self.crd_resource_handler.lightkube_client, crd_resources_manifests)
+        except ApiError as error:
+            # do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Failed to delete K8s resources, with error: {error}")
+                delete_error = error
 
-    def _check_image_details(self):
+        if delete_error is not None:
+            raise delete_error
```

Comment:
By catching both delete errors in the same variable, it means if both k8s and crd handlers raise errors we only find out about one.  Is that the intent here?  I like the idea of trying to delete as much as possible before raising here, but maybe catch the errors separately so both can be mentioned?

It gets a bit awkward since you can only raise one.  Maybe you could at least log about the other

Reply:
Maybe we could have a helper here, similar to my other suggestion on `_apply_k8s_resources()`, so you catch the errors and handle them individually. Your remove event handler could look like:

```python
def _on_remove(self, event):
  remove_helper(k8s_resource_handler)
  remove_helper(crd_resource_handler)
```

Reply:
Good idea, let me check.

Reply:
BTW, I can only raise one error, but I could have it include a list of Exceptions like this:
```python
delete_errors = []
for handler in {self.k8s_resource_handler, self.crd_resource_handler}:
    try:
        self._remove_k8s_resources(handler)
    except ApiError as error:
        delete_errors.append(error)

if delete_errors:
    raise GenericCharmRuntimeError("Removing K8s resources failed", delete_errors)
```
If we don't want to extend `GenericCharmRuntimeError` to accept `*args` we could just use a plain old `Exception` there (but let's go for the former).